network/p2p/inspector/internal/cache/cache_test.go

```diff
@@ -0,0 +1,549 @@
+func TestRecordCache_ConcurrentInitAndRemove(t *testing.T) {
+	cache := cacheFixture(t, 100, defaultDecay, zerolog.Nop(), metrics.NewNoopCollector())
+	require.NotNil(t, cache)
+	// expect cache to be initialized with a empty active cluster IDs list
+	require.Equalf(t, uint(0), cache.Size(), "cache size must be 1")
+
+	nodeIDs := unittest.IdentifierListFixture(20)
+	nodeIDsToAdd := nodeIDs[:10]
+	nodeIDsToRemove := nodeIDs[10:]
+
+	for _, nodeID := range nodeIDsToRemove {
+		cache.Init(nodeID)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(nodeIDs))
+
+	// initialize spam records concurrently
+	for _, nodeID := range nodeIDsToAdd {
+		go func(id flow.Identifier) {
+			defer wg.Done()
+			cache.Init(id)
+		}(nodeID)
+	}
+
+	// remove spam records concurrently
+	for _, nodeID := range nodeIDsToRemove {
+		go func(id flow.Identifier) {
+			defer wg.Done()
+			cache.Remove(id)
+			require.NotContains(t, id, cache.NodeIDs())
+		}(nodeID)
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+
+	// ensure that the initialized records are correctly added to the cache
+	// and removed records are correctly removed from the cache
+	require.Equal(t, uint(nodeIDsToAdd.Len()), cache.Size())
+}
+
+// TestRecordCache_ConcurrentInitRemoveUpdate tests the concurrent initialization, removal, and adjustment of
+// records for different node IDs. The test covers the following scenarios:
+// 1. Multiple goroutines initializing records for different node IDs concurrently.
+// 2. Multiple goroutines removing records for different node IDs concurrently.
+// 3. Multiple goroutines adjusting records for different node IDs concurrently.
+func TestRecordCache_ConcurrentInitRemoveUpdate(t *testing.T) {
+	cache := cacheFixture(t, 100, defaultDecay, zerolog.Nop(), metrics.NewNoopCollector())
+	require.NotNil(t, cache)
+	// expect cache to be initialized with a empty active cluster IDs list
+	require.Equalf(t, uint(0), cache.Size(), "cache size must be 1")
+
+	nodeIDs := unittest.IdentifierListFixture(30)
+	nodeIDsToAdd := nodeIDs[:10]
+	nodeIDsToRemove := nodeIDs[10:20]
+	nodeIDsToAdjust := nodeIDs[20:]
+
+	for _, nodeID := range nodeIDsToRemove {
+		cache.Init(nodeID)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(nodeIDs))
+
+	// Initialize spam records concurrently
+	for _, nodeID := range nodeIDsToAdd {
+		go func(id flow.Identifier) {
+			defer wg.Done()
+			cache.Init(id)
+		}(nodeID)
+	}
+
+	// Remove spam records concurrently
+	for _, nodeID := range nodeIDsToRemove {
+		go func(id flow.Identifier) {
+			defer wg.Done()
+			cache.Remove(id)
+			require.NotContains(t, id, cache.NodeIDs())
+		}(nodeID)
+	}
+
+	// Adjust spam records concurrently
+	for _, nodeID := range nodeIDsToAdjust {
+		go func(id flow.Identifier) {
+			defer wg.Done()
+			_, _ = cache.Update(id)
+		}(nodeID)
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
```

Comment:
This test might be missing an assertion about the cache state after all the concurrent updates complete. Ie. it should contain exactly `nodeIDsToAdjust + nodeIDsToAdd`